### DISCUSSION TITLE
MyPy "Attributes without a default cannot follow attributes with one"

### DISCUSSION THREAD

**Discussion by hubernhard** [sqlalchemy/sqlalchemy#10219] [Usage Questions]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

I have this setting with a `Base` class and an `AbstractBase` that adds additional columns which are only needed by some tables. Those tables inherit from `AbstractBase` and all others inherit from `Base` directly. 
The following example leads to MyPy complaining about "Attributes without a default cannot follow attributes with one". 

I understand that this is caused by the default value on `AbstractBase.a`. But if I omit the `AbstractBase` in between and add the additional column (including the default value) to the `Base` class directly, it works without an error.

Maybe I'm missing something?

Thank you for your help!

### To Reproduce

```python
from sqlalchemy import Identity
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    MappedAsDataclass,
    mapped_column,
)


class Base(MappedAsDataclass, DeclarativeBase, kw_only=True):
    id: Mapped[int | None] = mapped_column(
        Identity(), primary_key=True, default=None, nullable=False
    )


class AbstractBase(Base):
    __abstract__ = True

    a: Mapped[int] = mapped_column(default=0)


class Table(AbstractBase):
    __tablename__ = "table"

    b: Mapped[int]
```

### Error

```
test.py:25: error: Attributes without a default cannot follow attributes with one  [misc]
Found 1 error in 1 file (checked 1 source file)
``` 

### Versions

- OS: Windows
- Python: 3.11.4
- SQLAlchemy: 2.0.19
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): mypy 1.5.0

**Comment by CaselIT** [comment]
Hi,

This is just the default behaviour of dataclasses, and has nothing to do with how sqlalchemy is typed.
The most straight forward ways to solve are either don't use dataclasses or use kwonly in the attributes with default

**Reply by zzzeek** [reply]
this works

```py
from sqlalchemy import Identity
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    MappedAsDataclass,
    mapped_column,
)


class Base(MappedAsDataclass, DeclarativeBase, kw_only=True):
    id: Mapped[int | None] = mapped_column(
        Identity(), primary_key=True, default=None, nullable=False,
    )


class AbstractBase(Base):
    __abstract__ = True

    a: Mapped[int] = mapped_column(default=0, kw_only=True)


class Table(AbstractBase):
    __tablename__ = "table"

    b: Mapped[int]
```

**Reply by hubernhard** [reply]
Ok, thank you! I thought that the `kw_only=True` on `Base` gets passed down to `AbstractBase` already.

**Reply by zzzeek** [reply]
you could do this:

```py
class AbstractBase(Base, kw_only=True):
    __abstract__ = True

    a: Mapped[int] = mapped_column(default=0)

```


noting overall that these aren't our rules.  this is all the pep-681 rules baked into mypy

**Reply by hubernhard** [reply]
Ok my bad, I could have tried this as well 🙈  

Thank you for your quick response and for your work on sqlalchemy in general!

**Comment by blazing-gig** [comment]
> Ok, thank you! I thought that the `kw_only=True` on `Base` gets passed down to `AbstractBase` already.

@zzzeek Hey, in terms of functionality, isn't this the case ?

**Reply by zzzeek** [reply]
it's not supposed to be, if the type checkers don't expect that it is.   we are basically trying to match what pep-681 implementations think should be happening.
